cql3/restrictions/statement_restrictions.cc

```diff
@@ -969,5 +975,229 @@ bool single_column_restriction::LIKE::is_satisfied_by(bytes_view data, const que
     return (*_matcher)(data);
 }
 
+namespace wip {
+
+namespace {
+
+using cql3::selection::selection;
+
+/// Returns cdef's value from the fetched data.
+bytes_opt get_value(const column_definition* cdef, const selection& selection,
+                    const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+                    const std::vector<bytes_opt>& other_columns) {
+    switch (cdef->kind) {
+    case column_kind::partition_key:
+        return partition_key[cdef->id];
+    case column_kind::clustering_key:
+        return clustering_key[cdef->id];
+    case column_kind::static_column:
+    case column_kind::regular_column:
+        return other_columns[selection.index_of(*cdef)];
+    default:
+        throw exceptions::unsupported_operation_exception("Unknown column kind");
+    }
 }
+
+/// True iff columns' values equal t.
+bool equal(::shared_ptr<term> t, const idents& columns, const selection& selection,
+           const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+           const std::vector<bytes_opt>& other_columns, const query_options& options) {
+    if (auto multi = dynamic_pointer_cast<multi_item_terminal>(t)) {
+        const auto& rhs = multi->get_elements();
+        if (rhs.size() != columns.size()) {
+            throw std::logic_error("LHS and RHS size mismatch");
+        }
+        for (size_t i = 0; i < rhs.size(); ++i) {
+            if (rhs[i] != get_value(columns[i], selection, partition_key, clustering_key, other_columns)) {
+                return false;
+            }
+        }
+        return true;
+    } else  {
+        if (columns.size() != 1) {
+            throw std::logic_error("RHS for multi-column is not a tuple");
+        }
+        return to_bytes_opt(t->bind_and_get(options)) ==
+                get_value(columns[0], selection, partition_key, clustering_key, other_columns);
+    }
 }
+
+/// True iff lhs is limited by rhs in the manner prescribed by op.
+bool limits(const bytes& lhs, const operator_type& op, const bytes& rhs, const abstract_type& type) {
```

Comment:
bytes_view? or even bytes_view_opt? One of theses can be NULL.

Reply:
Neither can be null in this overload, see line 1091.

Reply:
What's the advantage of bytes_view over `const bytes&`?

Reply:
@dekimir In general - the ability to create bytes_view instances over any fragment of a bigger, non-owned buffer without any copying/allocations, + the ability to remove prefixes/suffixes at will, e.g. during parsing.

Reply:
Sure, but let me rephrase: what's the advantage of bytes_view for a parameter to which we always pass bytes?

Reply:
Hearing no good reason for `bytes_view`, I'm keeping this as is.

Reply:
Pardon for the delay - holidays :) I joined in the middle of the discussion, so I may lack context - I guess the only potential advantage would be future-proofness (is it a word?) in case we actually need `bytes_view` in another layer later - then, even a single `const bytes&` in a call-stack would mean having to copy the bytes, since one can't create `const bytes&` from `bytes_view` (but, conversely, one can create `bytes_view` from `const bytes&` if need be).

Reply:
Another minor advantage is that extracting begin/end pointers from bytes_view is faster than from const bytes&. For one, passing by value is through a register pair while by reference requires an extra indirection. For two, calculating begin() from const bytes& requires checking whether the data fits into the object or not. Of course, you have to do that when you convert from bytes to bytes_view, so if you only use the bytes_view once, there is no gain. But if you use it several times, you win (if you don't use it at all, bytes& is faster).

In general it's better to use the weakest type in the API so callers can decay if they have a stronger type, rather than construct if they have a weaker type.

Reply:
Switched to bytes_view.